rfcs/syntax-static-imports.md

```diff
@@ -0,0 +1,281 @@
+
+## Design
+
+This RFC proposes the addition of a static "import statement" to supersede almost all of the current use of `require()` in static contexts. *Almost* all, because there are many valid use cases for wanting to dynamically import modules, and these use cases are outside the scope of this change.
+
+```Lua
+!import "foo/bar/baz"
+```
+
+The `!import` used here is designed to mirror the current `export` used for exporting type annotations. To avoid ambiguity with function call syntax, the exclamation point is added, which lines up in style with the current `--!strict` style. This syntax may perhaps be extended in the future to cover other types of static declaration - perhaps even the aforementioned type checking mode.
+
+Unlike `require()` it is intended to introduce members to the namespace statically, not to return a value dynamically. This is why it is specifically a statement, not an expression:
+
+```Lua
+local foo = doSomething(require("foo")) -- ok
+local bar = doSomething(!import "foo") -- not ok
+```
+
+Since this statement is statically evaluated, the argument *must* be statically evaluatable (as is currently done to provide typechecking for `require()`):
+
+```Lua
+!import "foo" -- ok
+
+local bar = "foo"
+!import bar -- ok
+
+local bar = tostring(os.clock())
+!import bar -- not ok
+```
+
+The reason this is not specifically limited to string literals is to allow Roblox-like environments to evaluate statements for their imports, including the use of previously defined constant-like values:
+
+```Lua
+!import script.Parent.Libraries.Fusion
+
+local Package = script.Parent
+!import Package.Libraries.Fusion
+!import Package:FindFirstChild("Libraries"):FindFirstChild("Fusion")
+!import Package:WaitForChild("Libraries"):WaitForChild("Fusion")
+```
+
+### Syntax: Basic form
+
+The most basic form gives only the expression which resolves statically to the module path:
+
+```Lua
+!import "foo/bar/baz"
+```
+
+The returned value from the module is placed in a local variable, adopting the module's name. Types are placed in a namespace of the same name. This is how `require()` works today.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local baz = require("foo/bar/baz")
+```
+
+### Syntax: Renaming
+
+To use a different name for the local variable/type namespace, an equals sign is added at the start of the statement:
+
+```Lua
+!import not_baz = "foo/bar/baz"
+```
+
+Beyond user convenience, this allows code to deal with modules named identically, as can be done today with `require()`.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local not_baz = require("foo/bar/baz")
+```
+
+### Syntax: Type prefix
+
+To import only the types from a module, and skip evaluating the module at runtime, the `type` keyword can be added.
```

Comment:
I don't have the issue on hand but FWIW this has been ice boxed as something that's extremely complicated to implement.

Reply:
Noted. I will update the RFC.